nil/services/rpc/rawapi/client_direct_emulator.go

```diff
@@ -0,0 +1,104 @@
+package rawapi
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/NilFoundation/nil/nil/common/check"
+	"github.com/NilFoundation/nil/nil/common/logging"
+	"github.com/NilFoundation/nil/nil/internal/network"
+	"github.com/NilFoundation/nil/nil/internal/types"
+)
+
+func newShardApiClientDirectEmulator[
+	ShardApiType shardApiBase,
+	T interface {
+		~*S
+		shardApiRequestPerformerSetter
+	},
+	S any,
+](
+	apiName string,
+	shardApi ShardApiType,
+	apiType reflect.Type,
+	transportType reflect.Type,
+) (T, error) {
+	codec, err := newApiCodec(apiType, transportType)
+	if err != nil {
+		return nil, err
+	}
+
+	var rv T = new(S)
+	rv.setShardApiRequestPerformer(&shardApiRequestPerformerDirectEmulator[ShardApiType]{
+		apiName:       apiName,
+		apiType:       apiType,
+		transportType: transportType,
+		shardApi:      shardApi,
+		codec:         codec,
+		derived:       rv,
+	})
+	return rv, nil
+}
+
+type shardApiRequestPerformerDirectEmulator[RawApiType shardApiBase] struct {
+	apiName       string
+	apiType       reflect.Type
+	transportType reflect.Type
+
+	shardApi RawApiType
+	codec    apiCodec
+
+	derived any
```

Comment:
Is it possible to make it typed or at least put a comment what is it?

Reply:
I didn't manage to describe it using the limitations of Go generics right away. I was very happy to see that this field is passed to a function that just wants `any`. But I'll see what I can do, including a comment.